### PR TITLE
Remove and filter operations new "drop_axiom_annotation" option

### PR DESCRIPTION
**Pull request by hkir-dev**
Resolves [#886]

- [ ] `docs/` have been added/updated
- [ ] tests have been added/updated
- [ ] `mvn verify` says all tests pass
- [ ] `mvn site` says all JavaDocs correct
- [ ] `CHANGELOG.md` has been updated

New parameter `drop_axiom_annotation` added to `remove` and `filter` operations. An annotation property is provided to this option to drop/remove all axiom annotations using the property.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 - Add new command: [`expand`] [#964]
+- Add `--drop_axiom_annotation` option to drop axiom annotations in [`remove`] and [`filter`] [#1023]
```
Comment (matentzn) [inline CHANGELOG.md:14]: Always kebap case for robot long options; https://www.theserverside.com/definition/Kebab-case

File: docs/filter.md [hunk 2]
```diff
@@ -28,6 +28,8 @@ Then `filter --term A --term R --term C --select "self parents" --axioms all --s
     - the objects for `ax2` are `{A, R, C}` (with `R some C` excluded), and all of these are in the target set, so `ax2` is matched and copied
     - the objects for `ax3` are `{D, E}`, and none of these are in the target set, so `ax3` is not matched and is not copied
 
+Finally `--drop_axiom_annotation` option lets you to specify an annotation property to drop all axiom annotations using it.
```
Comment (matentzn) [inline docs/filter.md:31]: kebap

File: docs/filter.md [hunk 3]
```diff
@@ -94,3 +96,13 @@ Create a "base" subset that only includes internal axioms (alternatively, use `r
       --include-term IAO:0000119 \
       --output results/template-base-filter.owl
 
+Create a "base" subset that axiom annotations related with the two properties are removed:
```
Comment (matentzn) [inline docs/filter.md:99]: Create a "base" subset in which axiom annotations involving IAO:0000117 and IAO:0000119 are removed:

File: docs/remove.md [hunk 4]
```diff
@@ -49,6 +49,8 @@ Then `remove --term A --term R --term C --select "self parents" --axioms all --s
     - the objects for `ax1` are `{A, B}`, and at least one of these is in the target set, so `ax1` is matched and removed
     - the objects for `ax2` are `{A, R, C}` (with `R some C` excluded), and at least one of these is in the target set, so `ax2` is matched and removed
     - the objects for `ax3` are `{D, E}`, and none of these are in the target set, so `ax3` is not matched and is not removed
+    
+Finally `--drop_axiom_annotation` option lets you to specify an annotation property to drop all axiom annotations using it.
```
Comment (matentzn) [inline docs/remove.md:53]: kebap

File: docs/remove.md [hunk 5]
```diff
@@ -227,3 +229,12 @@ Create a "base" subset by removing external axioms (alternatively, use `filter -
       --select annotation-properties \
       --signature true \
       --output results/filter_annotations.owl
+
+Create a "base" subset that axiom annotations related with the two properties are removed:
```
Comment (matentzn) [inline docs/remove.md:233]: Create a "base" subset in which axiom annotations involving IAO:0000117 and IAO:0000119 are removed:

File: robot-command/src/main/java/org/obolibrary/robot/FilterCommand.java [hunk 6]
```diff
@@ -43,6 +43,8 @@ public FilterCommand() {
     o.addOption("r", "trim", true, "if true, keep axioms containing only selected objects");
     o.addOption(
         "S", "signature", true, "if true, keep axioms with any selected entity in their signature");
+    o.addOption(
+        "d", "drop_axiom_annotation", true, "annotation property to drop axiom annotations");
```
Comment (matentzn) [inline robot-command/src/main/java/org/obolibrary/robot/FilterCommand.java:47]: kebap

Comment (matentzn) [reply robot-command/src/main/java/org/obolibrary/robot/FilterCommand.java:47]: "annotation property to drop axiom annotations" -> "drop all axiom annotations involving a particular annotation property"

File: robot-command/src/main/java/org/obolibrary/robot/FilterCommand.java [hunk 7]
```diff
@@ -212,12 +216,40 @@ public CommandState execute(CommandState state, String[] args) throws Exception
...
+   * uses annotation properties specified by the drop_axiom_annotation argument.
+   *
+   * @param line CommandLine to get options from
+   * @param ioHelper IOHelper to get IRIs
+   * @param ontology OWLOntology to drop axiom annotations from
+   */
+  private void dropAxiomAnnotations(CommandLine line, IOHelper ioHelper, OWLOntology ontology) {
+    OWLOntologyManager manager = ontology.getOWLOntologyManager();
+    List<String> annotationsToDrop =
+        CommandLineHelper.getOptionalValues(line, "drop_axiom_annotation");
```
Comment (matentzn) [inline robot-command/src/main/java/org/obolibrary/robot/FilterCommand.java:239]: keb.

File: robot-command/src/main/java/org/obolibrary/robot/RemoveCommand.java [hunk 8]
```diff
@@ -198,12 +202,40 @@ public CommandState execute(CommandState state, String[] args) throws Exception
...
 
+  /**
+   * Given a command line and an IOHelper, drops axiom annotations from the given ontology which
+   * uses annotation properties specified by the drop_axiom_annotation argument.
+   *
+   * @param line CommandLine to get options from
+   * @param ioHelper IOHelper to get IRIs
+   * @param ontology OWLOntology to drop axiom annotations from
+   */
+  private void dropAxiomAnnotations(CommandLine line, IOHelper ioHelper, OWLOntology ontology) {
```
Comment (matentzn) [inline robot-command/src/main/java/org/obolibrary/robot/RemoveCommand.java:222]: Why isnt this method in OntologyUtils or some such? This should not be duplicated across both commands?

File: robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java [hunk 9]
```diff
@@ -2344,4 +2344,26 @@ private static void spanGapsHelper(
...
+
+  /**
+   * Removes all of the axiom annotations for the given annotation property.
+   *
+   * @param ontology OWLOntology to get axioms
+   * @param property Annotation property to remove related axiom annotations.
+   */
+  public static void removeAxiomAnnotations(OWLOntology ontology, OWLAnnotationProperty property) {
+    OWLOntologyManager manager = ontology.getOWLOntologyManager();
+    for (OWLAxiom axiom : ontology.getAxioms()) {
```
Comment (matentzn) [inline robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java:2356]: We dont have to specify IMPORTS.INCLUDED or IMPORTS.EXCLUDED here?

**Review by matentzn — CHANGES_REQUESTED** [review]
Excellent design to allow specifying the APs to remove.

One thing that would be great: can we add a blanket option for --drop-axiom-annotations (should be camel case, and should have an `s` in the end, because there can be multiple), that allows removing all axioms annotations? Like --drop-axiom-annotations `all`?

File: CHANGELOG.md (hunk 1 above)
Comment (hkir-dev) [reply CHANGELOG.md]: Fixed

File: docs/filter.md (hunk 2 above)
Comment (hkir-dev) [reply docs/filter.md]: Fixed

File: docs/remove.md (hunk 4 above)
Comment (hkir-dev) [reply docs/remove.md]: Fixed

File: docs/remove.md (hunk 5 above)
Comment (hkir-dev) [reply docs/remove.md]: Fixed

File: robot-command/src/main/java/org/obolibrary/robot/FilterCommand.java (hunk 6 above)
Comment (hkir-dev) [reply robot-command/src/main/java/org/obolibrary/robot/FilterCommand.java]: Fixed

File: robot-command/src/main/java/org/obolibrary/robot/FilterCommand.java (hunk 7 above)
Comment (hkir-dev) [reply robot-command/src/main/java/org/obolibrary/robot/FilterCommand.java]: fixed and method moved to `RelatedObjectsHelper.java`

File: robot-command/src/main/java/org/obolibrary/robot/RemoveCommand.java (hunk 8 above)
Comment (hkir-dev) [reply robot-command/src/main/java/org/obolibrary/robot/RemoveCommand.java]: generic methods moved to `OntologyHelper.java` and methods only used by the FilterCommand and RemoveCommand are moved to `RelatedObjectsHelper.java`

File: robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java (hunk 9 above)
Comment (hkir-dev) [reply robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java]: method moved to `OntologyHelper.java` and imports excluded

**Review by matentzn — APPROVED** [review]
The code looks great, let me summarise its current behaviour.

1. We introduce a new option `--drop-axiom-annotations` which removes all axiom annotations (or those using one or more specific properties) from a target set of axioms, determined by either the `filter` or `remove` commands. 
2. It works identically in `remove` and `filter`. This is cognitively much easier than applying "filter" logic in "filter" (i.e. only **including** the selected axiom annotations in the target set), which is correctly reserved to the --select/--axioms pipeline.

The only slight question mark I have, without wanting to cause an issue, is whether 

1. This should be a **command in its own right** after all `robot remove-axiom-annotations --select all`. Basically, we could then pipe the output of filter/remove to `remove-axiom-annotations` to achieve the same effect, without overloading the `remove` command too much.
2. Or, even if we go with the current implementation, we may not need to extend `filter`, and instead rely on passing the output of filter to `remove --drop-axiom-annotations all`. 

I will let @jamesaoverton think about this one, but I think this is the last remaining question here.

**Comment by matentzn** [comment]
The implementation for this PR is correct, and we can use it to do what we want. But I have some doubts about the cleanliness of the design, as described in my comment above. To get in the headspace again:

Remove/filter does its thing to the very end.... and then, the target set (aka whatever remains) gets post processed using the --drop-axiom-annotations option. This does overload the filter and axiom commands a little bit, as this is, indeed, a conceptually independent operation. On the other hand: practically, I cant think of a case where we would use this command independently of filter and remove (maybe with a tiny exception of "extract", but very corner case), and there is overhead in adding a new command - you are a better judge of the trade-offs here.

**Comment by jamesaoverton** [comment]
I realized that the integration tests were not running, so I made #1181 to fix that.
